app/projectsmodel.cpp

```diff
@@ -14,192 +8,582 @@
-    return;
+    proj->local = std::unique_ptr<LocalProject>( new LocalProject( project ) );
+    if ( proj->isMergin() )
+      proj->mergin->status = ProjectStatus::projectStatus( proj );
+
+    QModelIndex editIndex = index( mProjects.indexOf( proj ) );
+
+    emit dataChanged( editIndex, editIndex );
   }
+}
 
-  ProjectFile project = mProjectFiles.at( row );
+void ProjectsModel::onProjectDetachedFromMergin( const QString &projectFullName )
+{
+  std::shared_ptr<Project> proj = projectFromId( projectFullName );
 
-  mLocalProjects.deleteProjectDirectory( mLocalProjects.dataDir() + "/" + project.folderName );
+  if ( proj )
+  {
+    proj->mergin.reset();
+    QModelIndex editIndex = index( mProjects.indexOf( proj ) );
 
-  beginResetModel();
-  mProjectFiles.removeAt( row );
-  endResetModel();
+    emit dataChanged( editIndex, editIndex );
+
+    // This project should also be removed from project list for remote project model types,
+    // however, currently one needs to click on "My projects/Shared/Explore" and that sends
+    // another listProjects request. In new list this project will not be shown.
+  }
+}
+
+void ProjectsModel::onProjectAttachedToMergin( const QString &projectFullName )
+{
+  // To ensure project will be in sync with server, send listProjectByName request.
+  // In theory we could send that request only for this one project.
+  listProjectsByName();
+}
+
+void ProjectsModel::onAuthChanged()
+{
+  if ( !mBackend->userAuth() || !mBackend->userAuth()->hasAuthData() ) // user logged out, clear created and shared lists
+  {
+    if ( mModelType == CreatedProjectsModel || mModelType == SharedProjectsModel )
+    {
+      beginResetModel();
+      mProjects.clear();
+      endResetModel();
+    }
+  }
 }
 
-int ProjectModel::rowCount( const QModelIndex &parent ) const
+void ProjectsModel::setMerginApi( MerginApi *merginApi )
 {
-  Q_UNUSED( parent );
-  return mProjectFiles.count();
+  if ( !merginApi || mBackend == merginApi )
+    return;
+
+  mBackend = merginApi;
+  initializeProjectsModel();
 }
 
-QString ProjectModel::dataDir() const
+void ProjectsModel::setLocalProjectsManager( LocalProjectsManager *localProjectsManager )
 {
-  return mLocalProjects.dataDir();
+  if ( !localProjectsManager || mLocalProjectsManager == localProjectsManager )
+    return;
+
+  mLocalProjectsManager = localProjectsManager;
+  initializeProjectsModel();
 }
 
-QString ProjectModel::searchExpression() const
+void ProjectsModel::setModelType( ProjectsModel::ProjectModelTypes modelType )
 {
-  return mSearchExpression;
+  if ( mModelType == modelType )
+    return;
+
+  mModelType = modelType;
+  initializeProjectsModel();
 }
 
-void ProjectModel::setSearchExpression( const QString &searchExpression )
+QString ProjectsModel::modelTypeToFlag() const
 {
-  if ( searchExpression != mSearchExpression )
+  switch ( mModelType )
+  {
+    case CreatedProjectsModel:
+      return QStringLiteral( "created" );
```

Comment:
is this shown to user (translate?)

Reply:
No this is not, it is used to convert model type to listprojects compatible flag, e.g. created/shared